File: src/main/kotlin/ai/koalla/core/tools/EscalationTools.kt
package ai.koalla.core.tools

import ai.koalla.core.gateway.ChatwootGateway
import kotlinx.coroutines.runBlocking
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Description
import java.util.function.Function

/**
 * Spring AI function beans for human escalation.
 */
@Configuration
class EscalationTools(
    private val contextHolder: ToolContextHolder,
    private val chatwootGateway: ChatwootGateway
) {

    data class EscalateToHumanRequest(
        val summary: String = "",
        val lastMessage: String = ""
    )

    @Bean("escalateToHuman")
    @Description("""
        Escalate conversation to human agent immediately.
        Use when: user is dissatisfied, topic out of scope, user asks for human, or wants to stop messages.
        Parameters:
        - summary: Brief summary of conversation and why it needs human attention
        - lastMessage: User's last message
    """)
    fun escalateToHuman(): Function<EscalateToHumanRequest, String> = Function { req ->
        val ctx = contextHolder.require()
        runBlocking {
            val currentLabels = ctx.labels.toMutableSet()
            currentLabels.add("agente-off")
            chatwootGateway.updateLabels(ctx.accountId, ctx.conversationId, currentLabels.toList())

            val alertConvId = ctx.alertConversationId
            if (alertConvId.isNotBlank()) {
                val alertConvIdInt = alertConvId.toIntOrNull()
                if (alertConvIdInt != null) {
                    val alertMsg = """
                        |🚨 *Escalação humana*
                        |Nome: ${ctx.contactName}
                        |Resumo: ${req.summary}
                        |Última mensagem: ${req.lastMessage}
                    """.trimMargin()
                    chatwootGateway.sendMessage(ctx.accountId, alertConvIdInt, alertMsg)
                }
            }
        }
        "Conversation escalated. A human agent will take over shortly."
    }
}

File: legacy/app/tools/escalation_tools.py
"""Human escalation tool — mirrors n8n subworkflow 05."""
from langchain_core.tools import tool
from typing import Annotated
from legacy.app.services import chatwoot_client

_ctx: dict = {}


def set_context(ctx: dict) -> None:
    _ctx.update(ctx)


@tool
async def escalate_to_human(
    summary: Annotated[str, "A brief summary of the conversation and why it needs human attention."],
    last_message: Annotated[str, "The user's last message."],
) -> str:
    """
    Escalate the conversation to a human agent immediately.
    Use when: user is very dissatisfied, topic is out of scope,
    user asks to speak with a person, or asks to stop receiving messages.
    """
    account_id = _ctx.get("account_id")
    conversation_id = _ctx.get("conversation_id")
    alert_conv_id = _ctx.get("alert_conversation_id")
    name = _ctx.get("contact_name", "Usuário")

    # Add label agente-off to stop bot from responding
    current_labels = _ctx.get("labels", [])
    new_labels = list(set(current_labels + ["agente-off"]))
    await chatwoot_client.update_labels(account_id, conversation_id, new_labels)

    # Notify manager if alert conversation is configured
    if alert_conv_id:
        alert_msg = (
            f"🚨 *Escalação humana*\n"
            f"Nome: {name}\n"
            f"Resumo: {summary}\n"
            f"Última mensagem: {last_message}"
        )
        await chatwoot_client.send_message(account_id, int(alert_conv_id), alert_msg)

    return "Conversation escalated. A human agent will take over shortly."
